p2p/ipld/write_test.go

```diff
@@ -75,12 +83,126 @@ func TestPutBlock(t *testing.T) {
 	}
 }
 
+type preprocessingApp struct {
+	abci.BaseApplication
+}
+
+func (app *preprocessingApp) PreprocessTxs(
+	req abci.RequestPreprocessTxs) abci.ResponsePreprocessTxs {
+	time.Sleep(time.Second * 2)
+	randTxs := generateRandTxs(64, 256)
+	randMsgs := generateRandNamespacedRawData(128, nmt.DefaultNamespaceIDLen, 256)
+	randMessages := toMessageSlice(randMsgs)
+	return abci.ResponsePreprocessTxs{
+		Txs:      append(req.Txs, randTxs...),
+		Messages: &tmproto.Messages{MessagesList: randMessages},
+	}
+}
+func generateRandTxs(num int, size int) [][]byte {
+	randMsgs := generateRandNamespacedRawData(num, nmt.DefaultNamespaceIDLen, size)
+	for _, msg := range randMsgs {
+		copy(msg[:nmt.DefaultNamespaceIDLen], consts.TxNamespaceID)
+	}
+	return randMsgs
+}
+
+func toMessageSlice(msgs [][]byte) []*tmproto.Message {
+	res := make([]*tmproto.Message, len(msgs))
+	for i := 0; i < len(msgs); i++ {
+		res[i] = &tmproto.Message{NamespaceId: msgs[i][:nmt.DefaultNamespaceIDLen], Data: msgs[i][nmt.DefaultNamespaceIDLen:]}
+	}
+	return res
+}
+
+func TestDataAvailabilityHeaderRewriteBug(t *testing.T) {
```

Comment:
👍🏻

Reply:
> From experience I have in Golang: If I have an import cycle, it means I messed up proper logic segregation between packages architecturally, and simply extracting specific the piece of code causing the cycle is a simple hack, not a complete solution.

same. It's a "code smell" for sure. I think I'll leave it the way it is for now, but we could totally move the contents of the `wrapper` package to the `types` package. Besides a massive refactor, I don't think we have other better options. I'm certainly not attached to the current structure by any means, and am open to suggestions.